cmd/zilstat.in

```diff
@@ -0,0 +1,463 @@
+					if ctl.type != sysctl.CTLTYPE_NODE]
+				for s in k:
+					if not s or (s.name.find("zil") == -1 and \
+					    s.name.find("dataset_name") == -1):
+						continue
+					name, value = s.name, s.value
+					objid = re.findall(r'0x[0-9A-F]+', name[obj_start:], re.I)[0]
+					if objid not in kstat:
+						kstat[objid] = dict()
+					zil_start = len(file + ".objset-" + objid + ".")
+					kstat[objid][name[zil_start:]] = value \
+					    if (name.find("dataset_name")) else int(value)
+
+		except IOError as e:
+			if e.errno == errno.EPIPE:
+				sys.exit(1)
+
+	def kstat_process_str(k, objset = "GLOBAL", zil_start = 0):
+			global kstat
+			if not k:
+				sys.exit(1)
+			kstat[objset] = dict()
+			for s in k:
+				if not s or (s.name.find("zil") == -1 and \
+				    s.name.find("dataset_name") == -1):
+					continue
+				name, value = s.name, s.value
+				kstat[objset][name[zil_start:]] = value \
+				    if (name.find("dataset_name")) else int(value)
+
+elif sys.platform.startswith('linux'):
+	def kstat_update(pool = None, objid = None):
+		global kstat
+		try:
+			kstat = {}
+			if pool is None:
+				k = [line.strip() for line in open("/proc/spl/kstat/zfs/zil")]
+				kstat_process_str(k)
+			elif objid:
+				file = "/proc/spl/kstat/zfs/" + pool + "/objset-" + objid
+				k = [line.strip() for line in open(file)]
+				kstat_process_str(k, objid)
+			else:
+				objsets = os.listdir(f'/proc/spl/kstat/zfs/{pool}')
+				for objid in objsets:
+					if objid.find("objset-") == -1:
+						continue
+					file = "/proc/spl/kstat/zfs/" + pool + "/" + objid
+					k = [line.strip() for line in open(file)]
+					kstat_process_str(k, objid.replace("objset-", ""))
+		except IOError as e:
+			if e.errno == errno.EPIPE:
+				sys.exit(1)
+
+	def kstat_process_str(k, objset = "GLOBAL", zil_start = 0):
+			global kstat
+			if not k:
+				sys.exit(1)
+
+			kstat[objset] = dict()
+			for s in k:
+				if not s or (s.find("zil") == -1 and \
+				    s.find("dataset_name") == -1):
+					continue
+				name, unused, value = s.split()
+				kstat[objset][name] = value \
+				    if (name == "dataset_name") else int(value)
+
+def zil_process_kstat():
+	global curr, pool_name, dataset_name
+	curr.clear()
+	if gflag == True:
+		kstat_update()
+		zil_build_dict()
+	else:
+		try:
+			if pool_name:
+				kstat_update(pool_name)
+				zil_build_dict(pool_name)
+			elif dataset_name:
+				datasets = dataset_name.split(',')
+				for ds in datasets:
+					objid = subprocess.run(['zfs', 'list', '-Hpo', 'objsetid', \
+						ds], stdout=subprocess.PIPE, \
+						stderr=subprocess.DEVNULL).stdout.decode('utf-8').strip()
+					if not objid:
+						continue
+					pool = ds.split('/')[0]
+					objid = hex(int(objid))
+					kstat_update(pool, objid)
+					zil_build_dict(pool)
+			else:
+				pools = subprocess.run(['zpool', 'list', '-Hpo', 'name'], \
+				    stdout=subprocess.PIPE, \
+				    stderr=subprocess.DEVNULL).stdout.decode('utf-8').split()
```

Comment:
I'd prefer this to not run on every iteration either, but at least it is not specified by user and some new pool may get created.

Reply:
Exactly. The "-d" option can be used if a user just wants to monitor some specific dataset(s) where the subprocess command runs on the first iteration only.